### ISSUE TITLE
Choose runner based on branch

### ISSUE THREAD

**Issue by loafofpiecrust** [hercules-ci/hercules-ci-agent#414]
Hello! I just started digging into Hercules CI docs a few days ago to evaluate whether I can use it for one of my projects. I have two totally separate environments in AWS for staging and production. I'd like to deploy one EC2 instance to each environment running the Hercules CI agent, because I have resources in a private subnet in each environment that the runner needs to access.

Then, ideally, I'd like to have pushes to certain branches go to the staging instance, others go to the production instance. Is there support for that with Hercules? If not, how might Hercules otherwise work for me in this scenario? Reading through the docs, it looks like any instance with the same pool key could receive the job for a given commit from the repo?

**Comment by roberth** [comment]
> any instance with the same pool key could receive the job for a given commit.

That is correct. Also if you have multiple agents, evaluation runs on one of them (currently restricted to x86_64-linux) and the builds are distributed between them.


> how might Hercules otherwise work for me

Most have a build cluster (or node) that is distinct from their application environments; a separate deployment that isn't your application's staging or production. This keeps the CI simple.

I suppose you might consider having a staging environment for your agents themselves, to test out agent configuration updates and new releases. This is largely unnecessary because the agent is stateless, so if any problem does occur, you can revert.

But really I'm trying to guess your underlying requirements now. I should just ask, what are they?

If you prefer not to broadcast security requirements, you can email support@hercules-ci.com instead.

**Comment by loafofpiecrust** [comment]
Thanks for the response! Basically what I'm trying to do is give my CI access to other private resources in AWS. For example, I'd like to run database migrations on an RDS node in a private subnet from the CI runner. My staging and production have separate databases in separate accounts, so I can't give one runner access to both.

**Comment by roberth** [comment]
That makes sense.
You could implement this using `requiredFeatures`. This is a Nix feature that is also implemented by Hercules CI effects.
Here's how it would look for production.

 - Add `"env-production"` to `nix.systemFeatures` on the production agent
 - If you write your job definition using the [`herculesCI` attribute](https://docs.hercules-ci.com/hercules-ci-agent/evaluation/)  (agent >=0.9) it might look like
   ```nix
   herculesCI = { branch, ... }: {
     onPush.default.outputs = {
       # ... add builds here
       
       effects.production = runIf (branch == "production") (
         mkEffect {
           requiredFeatures = "env-production";
         }
       );
     };
   }
   ```
   Variations are possible, depending on your needs. For instance, you can extend the above with an `effects.staging` attr, so you always evaluate both. Alternatively you could have one `effects.deployment` attribute that behaves differently based on the branch.

This way the agents will collaborate on all the builds via your binary cache, and the effects will only run on the agent with matching features.

> My staging and production have separate databases in separate [AWS] accounts, so I don't think I can give one runner access to both.

Technically, you could set up VPC peering between accounts, but you'd lose the isolation, which may be good for security.
